### PR TITLE
Ignore broken disks when BE starts up

### LINKED ISSUE #3746: Ignore broken disks in the storage_data_path
**Describe the bug**
When BE starts up, it will check data paths in storage_data_path to make sure they are actually existed. 
In some situations, the data path is existed but for some reasons it is unavailable (broken disk). We aren't able to read or write into the broken data path after BE starts up.

**To Reproduce**
For example, Path: '/data1' is a broken disk, which can not be read or written. 
However, the path:'/data1' really exists in the OS.
Add Path:'/data1' to  storage_data_path in be.conf.
Set 'ignore_broken_disk = true' in be.conf
Start BE

**Expected behavior**
Path: '/data1' will not be ignored.

### PR DESCRIPTION
**Pull request by spaces-X**
Ignore **broken disks** from storage_root_path:

- add read and write check for each dir in storage_root_path and ignore broken disks.
- add AutoDeallocator.h to manage the lifecycle of allocated data by malloc or posix_memalign.
- rewrite DataDir::_read_and_write_test_file function.

### CODE REVIEW

**Comment by morningman** [comment]
Hi @spaces-X, Could you create an issue to describe what problem you try to solve?

**Comment by spaces-X** [comment]
> Hi @spaces-X, Could you create an issue to describe what problem you try to solve?

Done [issue](https://github.com/apache/incubator-doris/issues/3746)

File: be/src/olap/utils.cpp [hunk 1]
```diff
@@ -973,6 +979,108 @@ OLAPStatus copy_file(const string& src, const string& dest) {
...
+    }
+    LOG(WARNING) << "error when try to read and write temp file under the data path and return false. [path=" << path << "]";
+    return false;
+}
+
+bool check_dir_existed(const string& path) {
+    boost::filesystem::path p(path.c_str());
+
+    try {
+        if (boost::filesystem::exists(p)) {
```
Comment (morningman) [inline be/src/olap/utils.cpp:1071]: Just use:
`return boost::filesystem::exists(p);`

**Review by morningman — COMMENTED** [review]
Could you please update the document of this config `ignore_broken_disk`, both English and Chinese:

```
docs/zh-CN/administrator-guide/config/be_config.md
docs/en/administrator-guide/config/be_config.md
```

Thank you!

File: be/src/service/doris_main.cpp [hunk 2]
```diff
@@ -136,6 +136,25 @@ int main(int argc, char** argv) {
         LOG(FATAL) << "parse config storage path failed, path=" << doris::config::storage_root_path;
         exit(-1);
     }
+    auto it = paths.begin();
```
Comment (kangkaisen) [inline be/src/service/doris_main.cpp:139]: ignore_broken_disk has handled in `parse_conf_store_paths`.

File: be/src/util/auto_deallocator.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+//  {
+//      ...
+//      char *s = nullptr;
+//      posix_memalign((void **)s, 512, 4096);
+//      AutoDeallocator tmps(s);
+//      char *p = (char *)malloc(100*sizeof(char));
+//      AutoDeallocator tmpp(p);
+//      ...
+//  }
+    class AutoDeallocator {
```
Comment (kangkaisen) [inline be/src/util/auto_deallocator.h:35]: We don't need this class. std::unique_ptr and std::shared_ptr is enough

**Review by spaces-X — COMMENTED** [review]
> Could you please update the document of this config `ignore_broken_disk`, both English and Chinese:
> 
> ```
> docs/zh-CN/administrator-guide/config/be_config.md
> docs/en/administrator-guide/config/be_config.md
> ```
> 
> Thank you!

Done!

File: be/src/service/doris_main.cpp (hunk 2 above)
Comment (kangkaisen) [reply be/src/service/doris_main.cpp:139]: OK, I see. Thanks.

**Review by kangkaisen — DISMISSED** [review]
+1, LGTM

**Comment by spaces-X** [comment]
Could I merge this pr to master？
Or what should I do next?

**Comment by morningman** [comment]
> Could I merge this pr to master？
> Or what should I do next?

Hi, you don't need to do anything. This pr will be merged after being approved for a period of time,  
about a day, in case some other reviews have review comments.

**Comment by spaces-X** [comment]
> > Could I merge this pr to master？
> > Or what should I do next?
> 
> Hi, you don't need to do anything. This pr will be merged after being approved for a period of time,
> about a day, in case some other reviews have review comments.

Pull Request for my first time. 
Thanks a lot.

File: be/src/olap/utils.cpp [hunk 4]
```diff
@@ -973,6 +979,104 @@ OLAPStatus copy_file(const string& src, const string& dest) {
...
+    }
+    LOG(WARNING) << "error when try to read and write temp file under the data path and return false. [path=" << path << "]";
+    return false;
+}
+
+bool check_dir_existed(const string& path) {
+    boost::filesystem::path p(path.c_str());
+
+    try {
+        return boost::filesystem::exists(p);
```
Comment (chaoyli) [inline be/src/olap/utils.cpp:1071]: You can use check_exist function in util/file_utils.h.
Boots::filesystem::exist will be replace by util/file_utils.h.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/utils.cpp (hunk 4 above)
Comment (chaoyli) [reply be/src/olap/utils.cpp:1071]: This is related to this issue https://github.com/apache/incubator-doris/issues/2857

**Comment by morningman** [comment]
Compilation failed. Did you miss `util/auto_deallocator.h`?

**Comment by spaces-X** [comment]
> auto_deallocator

sorry, i forgot to remove that line

File: be/src/olap/utils.cpp (hunk 4 above)
Comment (spaces-X) [reply be/src/olap/utils.cpp]: done

**Comment by spaces-X** [comment]
> Compilation failed. Did you miss `util/auto_deallocator.h`?

Hi, i have removed the redundant `include` line, and rebuild it successfully in my machine. 
How to handle with the remaining requested changes created by reviewer 'chaoyli', which have already been changed.

**Comment by morningman** [comment]
Hi @chaoyli,  Any more comment？

**Review by chaoyli — APPROVED** [review]

**Pull request merged**